File: go/bigFact.go
import (
	"math/big"
	"fmt"
	"math"
	"runtime"
	"strconv"
	"time"
)

func main() {

	runtime.GOMAXPROCS(runtime.NumCPU() * 2)
	start := time.Now()

	a, b := int64(3000000000000), int64(10000000)

	x := numOfFactDigits(a)
	fmt.Printf("numberOfDigits(%v!) = %v\n", a, x)
	xLen := len(strconv.FormatInt(x, 10))

	y := func() *big.Int {
		fmt.Printf("%v! is being calculated...\n", b)
		result := compute(b)
		l := numOfFactDigits(b)
		return big.NewInt(0).Div(result, big.NewInt(0).Exp(big.NewInt(10), big.NewInt(l-int64(xLen)), nil))
	}()
	fmt.Printf("selectDigits(%v!, 0, %v) = %v\n", b, xLen, y)

	z := func() int {
		digits := big.NewInt(0).Add(big.NewInt(x), y).String()
		sum := 0
		for _, c := range digits {
			sum += int(c - '0')
		}
		return sum
	}()
	fmt.Printf("sumOfDigits(%v + %v) = %v\n", x, y, z)

	elapsed := time.Now()
	fmt.Println("time =", elapsed.Sub(start))
}

func compute(n int64) *big.Int {

	if n < 0 {
		panic(fmt.Sprintf("%v < 0", n))
	}
	if n < 2 {
		return big.NewInt(1)
	}

	return rangeProduct(1, n)
}

func rangeProduct(start, end int64) *big.Int {

	if end-start == 1 {
		return mulInt64(start, end)
	}
	if end-start == 2 {
		return mulBigInt(mulInt64(start, start+1), big.NewInt(end))
	}

	mid := (start + end) / 2
	p := make(chan *big.Int, 2)
	go func() { p <- rangeProduct(start, mid) }()
	go func() { p <- rangeProduct(mid+1, end) }()

	return mulBigInt(<-p, <-p)
}

func mulInt64(x, y int64) *big.Int {
	return mulBigInt(big.NewInt(x), big.NewInt(y))
}

func mulBigInt(x, y *big.Int) *big.Int {
	return big.NewInt(0).Mul(x, y)
}

// too slow for huge numbers
func numOfDigits(n *big.Int) int64 {
	c := int64(math.Log(2)/math.Log(10)*float64(n.BitLen())) + 1
	r := big.NewInt(0).Exp(big.NewInt(10), big.NewInt(c-1), nil).Cmp(n)
	if r > 0 {
		return c - 1
	}
	return c
}

func numOfFactDigits(n int64) int64 {
	return int64(math.Floor(logGamma(float64(n+1))/math.Log(10)) + 1)
}


// GammaLn implementation

// Lanczos coefficients
var lanczos = [...]float64{0.99999999999999709182,
	57.156235665862923517, -59.597960355475491248,
	14.136097974741747174, -0.49191381609762019978,
	0.33994649984811888699e-4, 0.46523628927048575665e-4,
	-0.98374475304879564677e-4, 0.15808870322491248884e-3,
	-0.21026444172410488319e-3, 0.21743961811521264320e-3,
	-0.16431810653676389022e-3, 0.84418223983852743293e-4,
	-0.26190838401581408670e-4, 0.36899182659531622704e-5,
}

// value of g constant in the Lanczos approximation
const g = float64(607) / float64(128)

func logGamma(n float64) float64 {

	if math.IsNaN(n) || n <= 0 {
		return math.NaN()
	}

	sum := float64(0)
	for i, v := range lanczos {
		if i == 0 {
			sum += v
		} else {
			sum += v / (n + float64(i))
		}
	}

	a, b := n+0.5, n+g+0.5
	halfLn2pi := 0.5 * math.Log(2*math.Pi)

	return a*math.Log(b) - b + halfLn2pi + math.Log(sum/n)
}

File: scala/BigFact.scala
import java.util.concurrent.{ForkJoinPool, RecursiveTask}

import scala.concurrent.duration.DurationDouble

/**
  * @author ahmad
  */
object BigFact {

  def main(args: Array[String]): Unit = {

    println("JIT Warm up ^_^")
    for (_ <- 1 to 20) {
      100000 !
    }

    val start = System.nanoTime

    val (a, b) = (3000000000000L, 10000000L)

    val x = numOfFactDigits(a)
    println(s"numberOfDigits($a!) = $x")
    val xLen = x.toString.length

    val y = {

      println(s"$b! is being calculated...")
      val result = b !

      // Stirling Approximation is a faster way to compute b!,
      // but in this case it's not accurate enough ;)
      // val result = stirlingApproximation(b.toInt)

      // GammaLn is faster too,
      // but similar to Stirling Approximation, it's not accurate enough.
      // val ln = GammaLn.compute(b + 1)
      // val result: BigInt = BigDecimal(Math.E).pow(BigDecimal(ln)).toBigInteger

      // we're sure that numOfDigits(b!) is lower than Int.MaxValue !!!
      val len = numOfFactDigits(b).intValue
      result / BigInt(10).pow(len - xLen)
    }
    println(s"selectDigits($b!, 0, $xLen) = $y")

    val z = (x + y).toString.toList.map(_.asDigit).sum
    println(s"sumOfDigits($x + $y) = $z")

    val elapsed = System.nanoTime
    println(f"time = ${(elapsed - start).nanos.toMillis}%,d ms")
  }

  implicit class LongOps(val n: Long) extends AnyVal {
    def ! : BigInt = compute(n)
  }

  def compute(n: Long): BigInt =
    if (n < 0) throw new IllegalArgumentException(n + " < 0")
    else if (n < 2) 1
    else {
      val fjp = new ForkJoinPool(Runtime.getRuntime.availableProcessors() * 2)
      val res = fjp.invoke(new RangeProduct(1, n))
      fjp.shutdown()
      res
    }

  final class RangeProduct(start: Long, end: Long) extends RecursiveTask[BigInt] {
    override def compute(): BigInt =
      end - start match {
        case 1 => BigInt(start) * BigInt(end)
        case 2 => BigInt(start) * BigInt(start + 1) * BigInt(end)
        case _ =>
          val mid = (start + end) / 2
          val left = new RangeProduct(start, mid).fork()
          val right = new RangeProduct(mid + 1, end).fork()
          left.join() * right.join()
      }
  }

  /**
    * too slow for huge numbers
    */
  def numOfDigits(n: BigInt): Long = {
    val c = (Math.log(2) / Math.log(10) * n.bitLength + 1).toInt
    val r = BigInt(10).pow(c - 1).compare(n)
    if (r > 0) c - 1 else c
  }

  def numOfFactDigits(n: Long): Long =
    (Math.floor(GammaLn.compute(n + 1) / Math.log(10d)) + 1).longValue

  def stirlingApproximation(n: Int): BigInt = {
    val left = Math.sqrt(2d * Math.PI * n)
    val right = BigDecimal(n).pow(n) * BigDecimal(Math.E).pow(-n)
    val mul = BigDecimal(left) * right
    mul.toBigInt
  }

}

object GammaLn {

  // Lanczos coefficients
  private val lanczos = Array(0.99999999999999709182d,
    57.156235665862923517d, -59.597960355475491248d,
    14.136097974741747174d, -0.49191381609762019978d,
    0.33994649984811888699e-4d, 0.46523628927048575665e-4d,
    -0.98374475304879564677e-4d, 0.15808870322491248884e-3d,
    -0.21026444172410488319e-3d, 0.21743961811521264320e-3d,
    -0.16431810653676389022e-3d, 0.84418223983852743293e-4d,
    -0.26190838401581408670e-4d, 0.36899182659531622704e-5d)

  // value of g constant in the Lanczos approximation
  private val g = 607d / 128d

  def compute(n: Double): Double = {

    if (n.isNaN || n <= 0d)
      return Double.NaN

    val sum = lanczos.indices.map {
      case 0 => lanczos(0)
      case i => lanczos(i) / (n + i)
    }.sum

    val (a, b) = (n + 0.5d, n + g + 0.5d)
    val halfLn2pi = 0.5d * Math.log(2 * Math.PI)

    a * Math.log(b) - b + halfLn2pi + Math.log(sum / n)
  }

}

